### PR TITLE
improved the overall code structure according to issue #23

### PR DESCRIPTION
**Pull request by sanskarmodi8**
As asked in the issue #23 , I have refractored the code accordingly.

### CODE REVIEW

**Comment by sanskarmodi8** [comment]
Hey @vasugr , kindly look into this Pull Request. I have done all the changes you asked for.

**Comment by vasugr** [comment]
Hey @sanskarmodi8 , thanks for working on it so quickly. I will take a look at the PR during the weekend

**Comment by sanskarmodi8** [comment]
> Hey @sanskarmodi8 , thanks for working on it so quickly. I will take a look at the PR during the weekend

Sure. Let me know if you need some changes.

File: weather.py [hunk 1]
```diff
@@ -1,98 +1,46 @@
...
+            return page.content
+        except requests.exceptions.RequestException as e:
+            print(f"Error fetching page content: {e}")
+            return None
+
+    def get_weather(self, city, lat, lon):
+        url = f"https://forecast.weather.gov/MapClick.php?lat={lat}&lon={lon}"
+        page_content = self.get_page_content(url)
+
+        if page_content is not None:
```
Comment (vasugr) [inline weather.py:26]: change the condition to 
`if page_content:`

File: weather.py [hunk 2]
```diff
@@ -1,98 +1,46 @@
...
-        return None
+class WeatherScraper:
+    def __init__(self, cities_file):
+        self.cities = self.load_cities(cities_file)
+
+    def load_cities(self, file_path):
+        with open(file_path, 'r') as file:
+            return json.load(file)
+
+    def get_page_content(self, url):
```
Comment (vasugr) [inline weather.py:13]: rename it to `_get_page_content` as it is a private method

Comment (vasugr) [reply weather.py:13]: rename all other private methods as well

File: weather.py [hunk 3]
```diff
@@ -1,98 +1,46 @@
...
+
+        if page_content is not None:
+            soup = BeautifulSoup(page_content, "html.parser")
+            div = soup.find('div', id="seven-day-forecast-container")
+            if div:
+                li = div.find('li', class_="forecast-tombstone")
+                if li:
+                    p = li.find('p', class_="short-desc")
+                    if p:
+                        return f"\t{city}: {p.get_text()}"
```
Comment (vasugr) [inline weather.py:34]: This can me moved to another method `_extract_weather_from_page()`  which returns the weather condition or `None` if not found

**Review by vasugr — COMMENTED** [review]
Good work! Have added a few comments.

File: weather.py [hunk 4]
```diff
@@ -1,98 +1,49 @@
...
-        page = requests.get(url)
-        page.raise_for_status()  # Raise HTTPError for bad responses
-        return page.content
-    except requests.exceptions.RequestException as e:
-        print(f"Error fetching page content: {e}")
+class WeatherScraper:
+    def __init__(self, cities_file):
+        self.cities = self.load_cities(cities_file)
+
+    def load_cities(self, file_path):
```
Comment (vasugr) [inline weather.py:9]: This is a static method, move it out of the class

File: weather.py [hunk 5]
```diff
@@ -1,98 +1,49 @@
...
         return None
 
-# Function to get weather information for a city
-def get_weather(city, lat, lon):
-    # Construct URL for the city
-    url = f"https://forecast.weather.gov/MapClick.php?lat={lat}&lon={lon}"
-    
-    # Fetch page content
-    page_content = get_page_content(url)
+    def get_weather(self, city, lat, lon):
```
Comment (vasugr) [inline weather.py:34]: Rename this private method

File: weather.py [hunk 6]
```diff
@@ -1,98 +1,49 @@
...
+        url = f"https://forecast.weather.gov/MapClick.php?lat={lat}&lon={lon}"
+        page_content = self._get_page_content(url)
+        return self._extract_weather_from_page(page_conetnt)
+        
+        return f"Failed to retrieve data for {city}. Please check your internet connection."
```
Comment (vasugr) [inline weather.py:39]: Code will never reach here

File: weather.py [hunk 7]
```diff
@@ -1,98 +1,49 @@
...
+        try:
+            page = requests.get(url)
+            page.raise_for_status()
+            return page.content
+        except requests.exceptions.RequestException as e:
+            print(f"Error fetching page content: {e}")
+            return None
+            
+    def _extract_weather_from_page(self, page_content):
+        if page_content:
```
Comment (vasugr) [inline weather.py:23]: do an early exit, 
```
if not page_content:
    return None
```

similarly in rest of the if conditions below

File: weather.py [hunk 8]
```diff
@@ -1,98 +1,49 @@
...
+            soup = BeautifulSoup(page_content, "html.parser")
+            div = soup.find('div', id="seven-day-forecast-container")
+            if div:
+                li = div.find('li', class_="forecast-tombstone")
+                if li:
+                    p = li.find('p', class_="short-desc")
+                    if p:
+                        return f"\t{city}: {p.get_text()}"
```
Comment (vasugr) [inline weather.py:31]: city variable is not defined, this method will error out.

Return only `p.get_text()` and don't print city name

File: weather.py [hunk 9]
```diff
@@ -1,98 +1,49 @@
...
-def get_weather(city, lat, lon):
-    # Construct URL for the city
-    url = f"https://forecast.weather.gov/MapClick.php?lat={lat}&lon={lon}"
-    
-    # Fetch page content
-    page_content = get_page_content(url)
+    def get_weather(self, city, lat, lon):
+        url = f"https://forecast.weather.gov/MapClick.php?lat={lat}&lon={lon}"
+        page_content = self._get_page_content(url)
+        return self._extract_weather_from_page(page_conetnt)
```
Comment (vasugr) [inline weather.py:37]: Here if the return value is none, then return error message or return weather condition (without city name)

**Review by vasugr — COMMENTED** [review]
Thanks for the quick turnaround, have added a few more comments

**Comment by sanskarmodi8** [comment]
> Thanks for the quick turnaround, have added a few more comments

I have updated the code accordingly. Hope it's correct now.

**Comment by vasugr** [comment]
Can you add more indentation and make the output look better

```
SAN FRANCISCO: Red Flag Warning

APPLE VALLEY: Sunny

```

image

**Comment by sanskarmodi8** [comment]
> Can you add more indentation and make the output look better
> 
> ```
> SAN FRANCISCO: Red Flag Warning
> 
> APPLE VALLEY: Sunny
> ```
> 
> image

Done the changes. Kindly review if some other changes also needs to be done.

**Review by vasugr — APPROVED** [review]
Looks great!

**Comment by sanskarmodi8** [comment]
> Looks great!

Thank you.

**Comment by ducdamchi** [comment]
Great work @sanskarmodi8 ! I just added some code to have the output nicely formatted with a pylsy table. Hopefully it gets merged soon!

**Comment by sanskarmodi8** [comment]
> Great work @sanskarmodi8 ! I just added some code to have the output nicely formatted with a pylsy table. Hopefully it gets merged soon!

Thank you :)

**Pull request merged**